metric_test.go

```diff
@@ -0,0 +1,282 @@
+		t.Fatal(err)
+	}
+
+	if tok != tokenStreamTag {
+		t.Error("expected stream tag separator")
+	}
+
+	if lit != "|ST" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	tok, lit, err = s.scanTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenStreamTag {
+		t.Error("expected stream tag separator")
+	}
+
+	if lit != "|ST" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	mtBuf := bytes.NewBufferString("|MT[blah:blah]")
+	s = newMetricScanner(mtBuf)
+	tok, lit, err = s.peekTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenMeasurementTag {
+		t.Error("expected measurement tag separator")
+	}
+
+	if lit != "|MT" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	tok, lit, err = s.scanTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenMeasurementTag {
+		t.Error("expected measurement tag separator")
+	}
+
+	if lit != "|MT" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+}
+
+func TestMetricParser(t *testing.T) {
+	var cases = []struct {
+		input string // input
+		numST int    // number of derived stream tags
+		numMT int    // number of derived measurement tags
+		lit   string // metric name literal
+	}{
+		{
+			input: "testing",
+			numST: 0,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: "testing*",
+			numST: 0,
+			numMT: 0,
+			lit:   "testing*",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			numST: 1,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: "t|esting|ST[blah:blah]",
+			numST: 1,
+			numMT: 0,
+			lit:   "t|esting",
+		},
+		{
+			input: "t|Sesting|ST[blah:blah]",
+			numST: 1,
+			numMT: 0,
+			lit:   "t|Sesting",
+		},
+		{
+			input: "tel|a|M|Ssting|ST[blah:blah]",
+			numST: 1,
+			numMT: 0,
+			lit:   "tel|a|M|Ssting",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			numST: 1,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah,blah1:blah,blah2:blah]",
+			numST: 3,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST[b"QUFB":blah,blah1:b"QkJCQgo=",b"YWFh":b"YmJi"]`,
+			numST: 3,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah:blah:blah]",
+			numST: 1,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah]|ST[blah:blah]",
+			numST: 2,
+			numMT: 0,
+			lit:   "testing",
+		},
+		{
+			input: "testing|MT[blah:blah]",
+			numST: 0,
+			numMT: 1,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah]|MT[blah:blah]",
+			numST: 1,
+			numMT: 1,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah]|MT[blah:blah]|ST[blah:blah]",
+			numST: 2,
+			numMT: 1,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST["blah:|ST[]":blah]|MT[blah:",]:|MTblah"]`,
+			numST: 1,
+			numMT: 1,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST["blah:|ST[]":blah]|MT[blah:",]:|MTblah"]`,
+			numST: 1,
+			numMT: 1,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST["b:|ST[]":b]|MT[b:",]:|MTb"]|ST[a:b]|MT[c:d]`,
```

Comment:
Please add a test for `testing|ST["quote\"slash\\":bar]` which is has tag `quote"slash\` with value 'bar'.

Reply:
I have now added support for escaped characters within quoted text in the metric name. I have added this test to the test suite as well.